src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml

```diff
@@ -1743,7 +1743,13 @@
     <Field Name="Id" Type="int"/>
     <Field Name="ForceCopyOfNullableValueType" Type="bool"/>
   </Node>
-  
+
+  <Node Name="BoundLoweredIsPatternExpression" Base="BoundExpression">
```

Comment:
> I am very uncomfortable with introducing an expression node that represents a control flow and doesn't encompass the control flow

Moving side-effects from SpillSequence to this node will cause double visit of those statements, because SpillSequence won't live past the spiller, but BoundLoweredIsPatternResultExpression should (to emit destination labels).
```cs
// we can't emit these in the emitter because statements require the stack to be empty
builder.AddStatements(loweredIsPattern.LoweredDagStatements);
// we still have to pass this on which includes LoweredDagStatements again
return builder.Update(loweredIsPattern); 
```
>trying to bring it back to a logical expression form

The goal is to avoid synthesizing the result temp in local rewriter. That alone will trigger lots of other optimizations and we get close enough to a binary expression. The lowered dag does resemble a logical expression but that's not what I aimed to do. Actually doing that would require rewritting the entire lowered dag, either in the original form or in the emitter.

- Rewrite `BoundDecisionDag` as a binary expression tree (using logical operators and ternaries) which is [non](https://github.com/dotnet/roslyn/pull/65845)-[trivial](https://github.com/dotnet/roslyn/pull/67018). 
- Emit `jumpToTrueLabelIfTrue <cond>` as `<cond>` if there's no more statements following, and so forth.

Note that this optimization applies even if the dag is not expressible as an expression at all, consider a try-finally block as part of enumerable list-patterns. The boolean result can still be pushed to the stack which will happen outside of try-finally.